docs/reference/edgeql/cardinality.rst

```diff
@@ -0,0 +1,168 @@
+
+.. code-block:: edgeql-repl
+
+  db> select len('hello');
+  {5}
+  db> select len({'hello', 'world'});
+  {5, 5}
+
+.. _ref_reference_cardinality_cartesian:
+
+Cartesian products
+^^^^^^^^^^^^^^^^^^
+
+In case of element-wise operations that
+accept multiple arguments, the operation is applied to a cartesian product
+cross-product of all the input sets.
+
+.. code-block:: edgeql-repl
+
+  db> select {'aaa', 'bbb'} ++ {'ccc', 'ddd'};
+  {'aaaccc', 'aaaddd', 'bbbccc', 'bbbddd'}
+  db> select {true, false} or {true, false};
+  {true, true, true, false}
+
+By extension, if any of the input sets are empty, the result of applying an
+element-wise function is also empty. In effect, when EdgeDB detects an empty
+set, it "short-circuits" and returns an empty set without applying the
+operation.
+
+.. code-block:: edgeql-repl
+
+  db> select {} ++ {'ccc', 'ddd'};
+  {}
+  db> select {} or {true, false};
+  {}
+
+.. note::
+
+  Certain functions and operators avoid this "short-circuit" behavior by
+  marking their inputs as :ref:`optional <ref_eql_sdl_functions_syntax>`. A
+  notable example of an operator with optional inputs is the :eql:op:`??
+  <COALESCE>` operator.
+
+  .. code-block:: edgeql-repl
+
+    db> select <str>{} ?? 'default';
+    {'default'}
+
+
+Per-input cardinality
+^^^^^^^^^^^^^^^^^^^^^
+
+Ultimately, the distinction between "aggregate vs element-wise" operations is
+a false one. Consider the :eql:op:`IN` operation.
+
+.. code-block:: edgeql-repl
+
+  db> select {1, 4} in {1, 2, 3};
+  {true, false}
+
+This operator takes two inputs. If it was "element-wise" we would expect the
+cardinality of the above operation to the cartesian product of the input
+cardinalities: ``2 x 3 = 6``. It it was aggregate, we'd expect a singleton
+output.
+
+Instead, the cardinality is ``2``. This operator is element-wise with respect
+to the first input and aggregate with respect to the second. The "element-wise
+vs aggregate" concept isn't determined on a per-function/per-operator basis;
+it determined on a per-input basis.
+
+
+Type qualifiers
+^^^^^^^^^^^^^^^
+
+When defining custom functions, all inputs  are element-wise by default. The
+``set of`` :ref:`type qualifier  <ref_sdl_function_typequal>` is used to
+designate an input as aggregate. The ``optional`` qualifier marks the input as
+optional; an operation will be executed is an optional input is empty or
+omitted, whereas passing an empty set for a "standard" (non-optional)
+element-wise input will always result in an empty set.
+
+
+Cardinality computation
+^^^^^^^^^^^^^^^^^^^^^^^
+
+To compute the final cardinality of a function/operator call, take the
+cardinality of each input and apply the following transformations, based on
+the type qualifier (or lack thereof) for each:
+
+.. code-block::
+
+  element-wise:  N -> N
+  optional:      N -> N == 0 ? 1 : N
+  aggregate:     N -> 1
+
+The cardinality of the resulting set is the product of the resulting numbers.
```

Comment:
This is an oversimplification. What this chart and rule will actually give you is the _number of function invocations_. The actual final cardinality of the result set is going to depend on the cardinality of each of those invocations. It so happens that a great many functions in the standard library return exactly 1 result per invocation (`count`, `len`, `in`, pretty much all of the arithmetic operators, etc.). However, some have *optional* results, like `json_get(<json>[0,1,2], '10')` will actually return `{}`. Others return *multiple* (declared as `set of`) results, like `array_unpack([0,1,2])` which will return `{0, 1, 2}`. So the cardinality of the result ultimately depends of what kind of function you're calling and what can it possibly return for the particular inputs.

You can also define custom functions that return some set for each given input:
```
edgedb> create function foo(x: str) -> set of str 
....... using (select r := array_unpack(str_split(x, '')) filter r in {'a', 'e', 'i', 'o', 'u'});
OK: CREATE FUNCTION
edgedb> select foo({'hello', 'world'});
{'e', 'o', 'o'}
edgedb> select foo({'qqq', 'www'});
{}
```
Currently we don't allow custom functions with aggregate arguments, but they certainly don't have to produce only a single result either.